### ISSUE TITLE
heap-buffer-overflow exiv2/src/image.cpp:1031 in Exiv2::Internal::binaryToString[abi:cxx11](unsigned char const*, unsigned long, unsigned long)

### ISSUE THREAD

**Issue by fgeek** [Exiv2/exiv2#209]
http://bugs.fi/media/afl/exiv2/2018-01-09-exiv2-crash-002.tiff
4be065595e4b81e876e32c9c4705f8313f896d43

```
=15745==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60300000d3d6 at pc 0x7fc0828a11dd bp 0x7ffeb52c1950 sp 0x7ffeb52c1948
READ of size 1 at 0x60300000d3d6 thread T0
    #0 0x7fc0828a11dc in Exiv2::Internal::binaryToString[abi:cxx11](unsigned char const*, unsigned long, unsigned long) /home/hsalo/src/exiv2/src/image.cpp:1031
    #1 0x7fc0828d3c70 in Exiv2::IptcData::printStructure(std::ostream&, unsigned char const*, unsigned long, unsigned int) /home/hsalo/src/exiv2/src/iptc.cpp:364
    #2 0x7fc0828b1441 in Exiv2::Image::printIFDStructure(Exiv2::BasicIo&, std::ostream&, Exiv2::PrintStructureOption, unsigned int, bool, char, int) /home/hsalo/src/exiv2/src/image.cpp:470
    #3 0x7fc0828b3602 in Exiv2::Image::printTiffStructure(Exiv2::BasicIo&, std::ostream&, Exiv2::PrintStructureOption, int, unsigned long) /home/hsalo/src/exiv2/src/image.cpp:533
    #4 0x7fc082b3deeb in Exiv2::TiffImage::printStructure(std::ostream&, Exiv2::PrintStructureOption, int) /home/hsalo/src/exiv2/src/tiffimage.cpp:344
    #5 0x7fc082b3b9c4 in Exiv2::TiffImage::readMetadata() /home/hsalo/src/exiv2/src/tiffimage.cpp:187
    #6 0x556bd42d8ad6 in Action::Print::printSummary() /home/hsalo/src/exiv2/src/actions.cpp:296
    #7 0x556bd42dd077 in Action::Print::run(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) /home/hsalo/src/exiv2/src/actions.cpp:242
    #8 0x556bd427387b in main /home/hsalo/src/exiv2/src/exiv2.cpp:166
    #9 0x7fc0819052b0 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x202b0)
    #10 0x556bd4274729 in _start (/home/hsalo/builds/exiv2/2018-01-08/bin/exiv2+0x12729)

0x60300000d3d6 is located 0 bytes to the right of 22-byte region [0x60300000d3c0,0x60300000d3d6)
allocated by thread T0 here:
    #0 0x7fc0837c0d70 in operator new[](unsigned long) (/usr/lib/x86_64-linux-gnu/libasan.so.3+0xc2d70)
    #1 0x7fc0828b1358 in Exiv2::Image::printIFDStructure(Exiv2::BasicIo&, std::ostream&, Exiv2::PrintStructureOption, unsigned int, bool, char, int) /home/hsalo/src/exiv2/src/image.cpp:467
    #2 0x7fc0828b3602 in Exiv2::Image::printTiffStructure(Exiv2::BasicIo&, std::ostream&, Exiv2::PrintStructureOption, int, unsigned long) /home/hsalo/src/exiv2/src/image.cpp:533
    #3 0x7fc082b3deeb in Exiv2::TiffImage::printStructure(std::ostream&, Exiv2::PrintStructureOption, int) /home/hsalo/src/exiv2/src/tiffimage.cpp:344
    #4 0x7fc082b3b9c4 in Exiv2::TiffImage::readMetadata() /home/hsalo/src/exiv2/src/tiffimage.cpp:187
    #5 0x556bd42d8ad6 in Action::Print::printSummary() /home/hsalo/src/exiv2/src/actions.cpp:296
    #6 0x556bd42dd077 in Action::Print::run(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) /home/hsalo/src/exiv2/src/actions.cpp:242
    #7 0x556bd427387b in main /home/hsalo/src/exiv2/src/exiv2.cpp:166
    #8 0x7fc0819052b0 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x202b0)

SUMMARY: AddressSanitizer: heap-buffer-overflow /home/hsalo/src/exiv2/src/image.cpp:1031 in Exiv2::Internal::binaryToString[abi:cxx11](unsigned char const*, unsigned long, unsigned long)
Shadow bytes around the buggy address:
  0x0c067fff9a20: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c067fff9a30: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c067fff9a40: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c067fff9a50: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c067fff9a60: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x0c067fff9a70: fa fa fa fa fa fa fa fa 00 00[06]fa fa fa fd fd
  0x0c067fff9a80: fd fd fa fa fd fd fd fa fa fa fd fd fd fa fa fa
  0x0c067fff9a90: fd fd fd fa fa fa fd fd fd fa fa fa fd fd fd fa
  0x0c067fff9aa0: fa fa 00 00 00 07 fa fa 00 00 00 07 fa fa 00 00
  0x0c067fff9ab0: 00 07 fa fa 00 00 00 07 fa fa 00 00 00 07 fa fa
  0x0c067fff9ac0: 00 00 00 07 fa fa 00 00 00 07 fa fa 00 00 00 07
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==15745==ABORTING
```

**Comment by D4N** [comment]
Just a note concerning this bug report: although it'll probably go away by removing 
`printStructure`, the actual problem will stay and we should fix it anyway. 

I tried the reproducer a few weeks ago and iirc it was an out of bounds read due to the array length being calculated incorrectly.

**Comment by piponazo** [comment]
@fgeek I tried to download the to reproduce the issue but the link seems to be broken. Could you upload it here?

**Comment by fgeek** [comment]
@piponazo Heh. Domain name reflects well about the availability. I need to fix my DNS. Broken service reflects my busyness lately ;)

Here it is also attached:

[2018-01-09-exiv2-crash-002.tiff.zip](https://github.com/Exiv2/exiv2/files/1896102/2018-01-09-exiv2-crash-002.tiff.zip)

**Comment by piponazo** [comment]
Thanks! I will take a look to this issue

**Comment by piponazo** [comment]
With the current version of master, we are throwing an exception when detecting an invalid type value:

```
$ bin/exiv2.exe -pa D:/2018-01-09-exiv2-crash-002.tiff
invalid type value detected in Image::printIFDStructure:  40448
Exiv2 exception in print action for file D:/2018-01-09-exiv2-crash-002.tiff:
invalid type value detected in Image::printIFDStructure
```
I did not try to run it with the AddressSanitizer, but I suspect the problem might be fixed in a recent PR.

**Comment by fgeek** [comment]
@piponazo I can still reproduce this with c80b1b9d51689692bc865f1a1d16bf7fd2a532c4. I can also test with PR. Could you give me ID?

**Comment by D4N** [comment]
This got fixed by #461.
